lib/internal/util.js

```diff
@@ -217,3 +219,62 @@ module.exports = exports = {
   // default isEncoding implementation, just in case userland overrides it.
   kIsEncodingSymbol: Symbol('node.isEncoding')
 };
+
+const kCustomPromisifiedSymbol = Symbol('util.promisify.custom');
+const kCustomPromisifyArgsSymbol = Symbol('customPromisifyArgs');
+
+function promisify(orig) {
+  if (typeof orig !== 'function') {
+    const errors = require('internal/errors');
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'original', 'function');
+  }
+
+  if (orig[kCustomPromisifiedSymbol]) {
+    const fn = orig[kCustomPromisifiedSymbol];
+    if (typeof fn !== 'function') {
+      throw new TypeError('The [util.promisify.custom] property must be ' +
+                          'a function');
+    }
+    Object.defineProperty(fn, kCustomPromisifiedSymbol, {
+      value: fn, enumerable: false, writable: false, configurable: true
+    });
+    return fn;
+  }
+
+  // Names to create an object from in case the callback receives multiple
+  // arguments, e.g. ['stdout', 'stderr'] for child_process.exec.
+  const argumentNames = orig[kCustomPromisifyArgsSymbol];
+
+  function fn(...args) {
+    const promise = createPromise();
+    try {
+      orig.call(this, ...args, (err, ...values) => {
+        if (err) {
+          promiseReject(promise, err);
+        } else if (argumentNames !== undefined && values.length > 1) {
+          const obj = {};
+          for (var i = 0; i < argumentNames.length; i++)
+            obj[argumentNames[i]] = values[i];
+          promiseResolve(promise, obj);
+        } else {
+          promiseResolve(promise, values[0]);
+        }
+      });
+    } catch (err) {
```

Comment:
This catch may accidentally catch errors thrown from promiseReject if orig.call happened to be synchronous. E.g. the value was cached in memory and didn't need a remote lookup. promiseReject will be a no-op in that case.

Ideally if the execution hit line 253 and then 264, it should re-throw the error instead of no-oping. At least it could log a warning.